### PR TITLE
More controlled db reconnect

### PR DESCRIPTION
**Pull request by olkorhon**
## Status
**READY**

## Description
This pr removes own db reconnect implementations which collide with autoReconnect that mongoose provides. Provides more controlled handling of database connection outage.

## Impacted Areas in Application
* app/db.js

### CODE REVIEW

File: app/db.js [hunk 1]
```diff
@@ -7,42 +7,48 @@ const nconf = require('../config');
...
-      socketOptions: {keepAlive: 1},
-      auto_reconnect: true
-    }
-  }; */
   const options = {
     useMongoClient: true,
-    // logger: logger,
-    loggerLevel: 'warning' // @todo fetch from config file
+    logger: logger.log.bind(logger),
+    loggerLevel: 'debug' // @todo fetch from config file
```
Comment (jupe) [inline app/db.js]: maybe debug log is too much..?

File: app/db.js [hunk 2]
```diff
@@ -7,42 +7,48 @@ const nconf = require('../config');
...
   };
-  logger.info(`Create MongoDB connection: ${dbUrl}`);
+
+  // Resolve after database connection is ready and listening
+  const connectionPending = new Promise((resolve) => {
+    mongoose.connection.once('connected', resolve);
```
Comment (jupe) [inline app/db.js]: if this doesn't happen ever we kind of lose this promise..

File: app/db.js (hunk 1 above)
Comment (olkorhon) [reply app/db.js]: perhaps.. I'll change it to silly.

File: app/db.js (hunk 2 above)
Comment (olkorhon) [reply app/db.js]: Well it needs to happen at some point or there is not much we can do, maybe we could use Bluebirds promise timeout to automatically reject it if it doesn't resolve in acceptable time?

File: app/db.js (hunk 1 above)
Comment (jupe) [reply app/db.js]: silly? more logs? 🙄

File: app/db.js [hunk 3]
```diff
@@ -7,42 +7,48 @@ const nconf = require('../config');
...
+
+  logger.info(`Connecting to MongoDB: ${dbUrl}`);
   return mongoose
     .connect(dbUrl, options)
-    .then(() => {
-      mongoose.connection.on('error', () => {
-        logger.error(`Could not connect to MongoDB: ${dbUrl}`);
-      });
-    });
+    .then(() => connectionPending);
```
Comment (jupe) [inline app/db.js]: is this needed at all? doesn't  mongoose connect resolves when connected..?

File: app/db.js (hunk 2 above)
Comment (jupe) [reply app/db.js]: it still doesn't stop listening event and then it calls already solved promises..

File: app/db.js (hunk 3 above)
Comment (olkorhon) [reply app/db.js]: Looking at the documentation you are quite correct. It is not needed.

File: app/db.js [hunk 4]
```diff
@@ -7,42 +7,42 @@ const nconf = require('../config');
...
-      auto_reconnect: true
-    }
-  }; */
   const options = {
     useMongoClient: true,
-    // logger: logger,
-    loggerLevel: 'warning' // @todo fetch from config file
+    logger: logger.log.bind(logger),
+    loggerLevel: 'silly' // @todo fetch from config file
   };
```
Comment (jupe) [inline app/db.js]: how much this produces logs now? 🙄

File: app/db.js [hunk 5]
```diff
@@ -7,42 +7,42 @@ const nconf = require('../config');
...
   return close();
 }
 
+mongoose.connection.on('error', (error) => {
+  logger.error(`MongoDB connection error: ${error.message}`);
+});
+
 mongoose.connection.on('disconnected', () => {
-  logger.warn('MongoDB connection lost, try again');
-  connect();
```
Comment (jupe) [inline app/db.js]: so mongoose by default reconnects?

File: app/tools/logger.js [hunk 6]
```diff
@@ -65,7 +65,7 @@ class MasterLogger {
     const args = data['args'] || [];
     args.unshift(`<Worker#${worker.id}>`);
     try {
-      this.logger[level](...args);
+      this.logger.log(level, ...args);
```
Comment (jupe) [inline app/tools/logger.js]: is this method supported also clustered loggers?

File: app/db.js (hunk 5 above)
Comment (olkorhon) [reply app/db.js]: That's the image I got from the documentation, seemed to work when I manually crashed and brought up the mongo process a few times.

**Pull request merged**